src/libtools/graphalgorithm.cc

```diff
@@ -48,77 +49,265 @@ vector<Edge> edgeSetToVector_(set<Edge> edges) {
  * Public Functions *
  ********************/
 
-bool singleNetwork(Graph g, GraphVisitor& gv) {
-  exploreGraph(g, gv);
-  return gv.getExploredVertices().size() == g.getVertices().size() &&
-         g.getIsolatedNodes().size() == 0;
+bool singleNetwork(Graph graph, GraphVisitor& graph_visitor) {
+  exploreGraph(graph, graph_visitor);
+  return graph_visitor.getExploredVertices().size() ==
+             graph.getVertices().size() &&
+         graph.getIsolatedNodes().size() == 0;
 }
 
-vector<shared_ptr<Graph>> decoupleIsolatedSubGraphs(Graph g) {
+std::set<Edge> exploreBranch(Graph g, int starting_vertex, Edge edge) {
+  // Check if the starting vertex is in the graph
+  if (!g.vertexExist(starting_vertex)) {
+    throw invalid_argument(
+        "Cannot explore a branch of the graph when the "
+        "exploration is started from a vertex that does not exist within the "
+        "graph.");
+  }
+  if (!g.edgeExist(edge)) {
+    throw invalid_argument(
+        "Edge does not exist in the graph so exploration of"
+        " the graph branch cannot continue");
+  }
+  if (!edge.contains(starting_vertex)) {
+    throw invalid_argument(
+        "The edge determining which branch to explore does "
+        "not contain the starting vertex.");
+  }
+
+  set<Edge> branch_edges;
+  if (edge.getEndPoint1() == edge.getEndPoint2()) {
+    branch_edges.insert(edge);
+    return branch_edges;
+  }
+  Graph_BF_Visitor gv_breadth_first;
+  GraphNode gn;
+  pair<int, GraphNode> pr_gn(starting_vertex, gn);
+  gv_breadth_first.exploreNode(pr_gn, g);
+  gv_breadth_first.setStartingVertex(edge.getOtherEndPoint(starting_vertex));
+  gv_breadth_first.initialize(g);
+
+  branch_edges.insert(edge);
+  while (!gv_breadth_first.queEmpty()) {
+    Edge ed = gv_breadth_first.nextEdge(g);
+    branch_edges.insert(ed);
+    gv_breadth_first.exec(g, ed);
+  }
+
+  vector<Edge> neigh_edges = g.getNeighEdges(starting_vertex);
+  for (auto ed : neigh_edges) {
+    int neigh_vertex = ed.getOtherEndPoint(starting_vertex);
+    if (neigh_vertex != starting_vertex) {
+      if (gv_breadth_first.vertexExplored(neigh_vertex)) {
+        branch_edges.insert(ed);
+      }
+    }
+  }
+
+  return branch_edges;
+}
+/**
+ * \brief This class is to help keep track of which vertices have and have not
+ * been explored.
+ *
+ **/
+class ExplorationRecord {
```

Comment:
a class in an .cc file?

Reply:
Yes, the class is not needed outside of the source file, it just there to help clean up the source code. I wanted to hide it because I don't want it to be accessible outside of the source file.

Reply:
make it private in the header

Reply:
There are no classes in the header just functions. Plus the header is by nature public. I will argue with you on this point, this class should not go in the header. If I put it in the header, I'm saying that it is ok for someone else to use it. If I do that then I need to add better bounds checking etc...

Reply:
Comment added.